### PR TITLE
chore: enable eslint:no-shadow

### PR DESCRIPTION
**Pull request by straker**
This came up in [a recent pr](https://github.com/dequelabs/axe-core/pull/4086#discussion_r1261194989) so went ahead and did it (hurray for resolving some 111 issues). For every file I touched I did the following:

* Moved default export to the top
* Converted to es6 arrow functions and `const` / `let`
* Refactored some improper uses of `reduce` to better array functions (when the reduce caused the name conflict)
* Removed unneeded `/* global */` eslint strings

In some files that had naming conflict issues the issues were in nested functions so I moved the function outside the nested scope.

### CODE REVIEW

File: lib/checks/aria/aria-errormessage-evaluate.js [hunk 1]
```diff
@@ -24,10 +24,10 @@ import { isVisibleToScreenReaders } from '../../commons/dom';
  * @memberof checks
  * @return {Mixed} True if aria-errormessage references an existing element that uses a supported technique. Undefined if it does not reference an existing element. False otherwise.
  */
-function ariaErrormessageEvaluate(node, options, virtualNode) {
+export default function ariaErrormessageEvaluate(node, options, virtualNode) {
```
Comment (stephenmathieson) [inline lib/checks/aria/aria-errormessage-evaluate.js:27]: What is the motivation behind this change?

File: lib/checks/tables/td-headers-attr-evaluate.js [hunk 2]
```diff
@@ -14,12 +14,9 @@ function tdHeadersAttrEvaluate(node) {
...
 
-  const ids = cells.reduce((ids, cell) => {
-    if (cell.getAttribute('id')) {
-      ids.push(cell.getAttribute('id'));
-    }
-    return ids;
-  }, []);
+  const ids = cells
+    .filter(cell => cell.getAttribute('id'))
+    .map(cell => cell.getAttribute('id'));
```
Comment (stephenmathieson) [inline lib/checks/tables/td-headers-attr-evaluate.js:19]: This is much easier to read/understand 👍

File: lib/core/base/audit.js [hunk 3]
```diff
@@ -672,6 +520,158 @@ class Audit {
...
+  if (
+    window.location &&
+    window.location.origin &&
+    window.location.origin !== 'null'
+  ) {
+    return window.location.origin;
+  }
+}
+
+/*eslint no-unused-vars: 0*/
```
Comment (stephenmathieson) [inline lib/core/base/audit.js:540]: It doesn't look like this directive is necessary.

```suggestion
```

File: lib/core/base/audit.js [hunk 4]
```diff
@@ -672,6 +520,158 @@ class Audit {
...
+function getDefaultConfiguration(audit) {
+  var config;
```
Comment (stephenmathieson) [inline lib/core/base/audit.js:542]: nit-

```suggestion
  let config;
```

File: lib/core/base/audit.js [hunk 5]
```diff
@@ -672,6 +520,158 @@ class Audit {
...
+  }
+
+  config.rules = config.rules || [];
+  config.checks = config.checks || [];
+  config.data = { checks: {}, rules: {}, ...config.data };
+  return config;
+}
+
+function unpackToObject(collection, audit, method) {
+  var i, l;
```
Comment (stephenmathieson) [inline lib/core/base/audit.js:568]:
```suggestion
  let i, l;
```

File: lib/core/base/audit.js [hunk 6]
```diff
@@ -672,6 +520,158 @@ class Audit {
...
+ * Locale `b` follows the `axe.CheckLocale` shape and resembles:
+ *
+ *    {
+ *      pass: string,
+ *      fail: string,
+ *      incomplete: string | { [key: string]: string }
+ *    }
+ */
+
+const mergeCheckLocale = (a, b) => {
```
Comment (stephenmathieson) [inline lib/core/base/audit.js:599]: Just curious - when is a function declaration preferred over a function expression? Also why is one preferred over the other?

File: lib/core/reporters/helpers/process-aggregate.js [hunk 7]
```diff
@@ -55,7 +25,7 @@ var resultKeys = constants.resultGroups;
  * @return {Object}
  *
  */
-function processAggregate(results, options) {
+export default function processAggregate(results, options) {
   var resultObject = axe.utils.aggregateResult(results);
```
Comment (stephenmathieson) [inline lib/core/reporters/helpers/process-aggregate.js:29]:
```suggestion
  const resultObject = axe.utils.aggregateResult(results);
```

File: lib/core/reporters/helpers/process-aggregate.js [hunk 8]
```diff
@@ -107,4 +77,32 @@ function processAggregate(results, options) {
...
+function normalizeRelatedNodes(node, options) {
+  ['any', 'all', 'none'].forEach(type => {
+    if (!Array.isArray(node[type])) {
+      return;
+    }
+    node[type]
+      .filter(checkRes => Array.isArray(checkRes.relatedNodes))
+      .forEach(checkRes => {
+        checkRes.relatedNodes = checkRes.relatedNodes.map(relatedNode => {
+          var res = {
```
Comment (stephenmathieson) [inline lib/core/reporters/helpers/process-aggregate.js:89]:
```suggestion
          const res = {
```

File: lib/core/utils/get-flattened-tree.js [hunk 9]
```diff
@@ -67,8 +97,8 @@ function createNode(node, parent, shadowId) {
 function flattenTree(node, shadowId, parent) {
   // using a closure here and therefore cannot easily refactor toreduce the statements
   var retVal, realArray, nodeName;
-  function reduceShadowDOM(res, child, parent) {
-    var replacements = flattenTree(child, shadowId, parent);
+  function reduceShadowDOM(res, child, parentVNode) {
+    var replacements = flattenTree(child, shadowId, parentVNode);
```
Comment (stephenmathieson) [inline lib/core/utils/get-flattened-tree.js:101]:
```suggestion
    const replacements = flattenTree(child, shadowId, parentVNode);
```

File: lib/core/utils/get-rule.js [hunk 10]
```diff
@@ -3,15 +3,13 @@
  * @param {String} ruelId the rule id
  * @return {Rule}
  */
-function getRule(ruleId) {
+export default function getRule(ruleId) {
   // TODO: es-modules_audit
-  var rule = axe._audit.rules.find(rule => rule.id === ruleId);
+  var rule = axe._audit.rules.find(r => r.id === ruleId);
```
Comment (stephenmathieson) [inline lib/core/utils/get-rule.js:8]:
```suggestion
  const rule = axe._audit.rules.find(r => r.id === ruleId);
```

File: lib/checks/aria/aria-errormessage-evaluate.js (hunk 1 above)
Comment (straker) [reply lib/checks/aria/aria-errormessage-evaluate.js:27]: Follows our coding guidelines.

> If you encounter any code that we maintain that does not put the default export at the top, you should update the file to do so.

File: lib/core/base/audit.js (hunk 6 above)
Comment (straker) [reply lib/core/base/audit.js:599]: As far as I know we have no standard for when to use either. Mostly it's been up to the person writing the code.

File: test/integration/adapter.js [hunk 11]
```diff
@@ -1,5 +1,14 @@
 /*global mocha */
 var failedTests = [];
+var flattenTitles = function (test) {
+  var titles = [];
```
Comment (WilcoFiers) [inline test/integration/adapter.js:4]: While you're at it:

```suggestion
const failedTests = [];
function flattenTitles (test) {
  const titles = [];
```

**Review by WilcoFiers — DISMISSED** [review]
Couple suggestions.

File: lib/checks/aria/aria-errormessage-evaluate.js (hunk 1 above)
Comment (stephenmathieson) [reply lib/checks/aria/aria-errormessage-evaluate.js:27]: What is the motivation behind putting the default export at the top?

File: lib/core/base/audit.js (hunk 6 above)
Comment (stephenmathieson) [reply lib/core/base/audit.js:598]: Both functions created using expressions and declarations have a name.

```
Welcome to Node.js v16.14.0.
Type ".help" for more information.
> function foo(){}
undefined
> foo.name
'foo'
> const bar = () => {}
undefined
> bar.name
'bar'
>
```

File: test/checks/color/link-in-text-block-style.js [hunk 12]
```diff
@@ -1,67 +1,64 @@
...
   function createStyleString(selector, outerStyle) {
     // Merge style with the default
-    var prop;
-    var styleObj = {};
-    for (prop in defaultStyle) {
+    const styleObj = {};
+    for (const prop in defaultStyle) {
       if (defaultStyle.hasOwnProperty(prop)) {
         styleObj[prop] = defaultStyle[prop];
       }
```
[the remark this answers is not in the document]
Comment (straker) [reply test/checks/color/link-in-text-block-style.js:43]: `defaultStyle` and `outerStyle` are both objects so this won't work. Did you mean to maybe iterate over the keys?

```js
for (const prop of Object.keys(defaultStyle)) {
  styleObj[prop] = defaultStyle[prop];
}
```

[the remark this answers is not in the document]
Comment (WilcoFiers) [reply test/checks/color/link-in-text-block-style.js:43]: 🤦 Oh right, never mind!

File: lib/checks/aria/aria-errormessage-evaluate.js (hunk 1 above)
Comment (straker) [reply lib/checks/aria/aria-errormessage-evaluate.js:27]: Ensures that when you open the file the main code path is the first thing you read.

**Review by WilcoFiers — APPROVED** [review]
